zig/shell.nix

```diff
@@ -0,0 +1,67 @@
+{ pkgs ? import <nixpkgs> {} } :
+let
+	# Right. Getting this to work took more effort than all the other shells put together.
+	# Things that didn't work:
+	# - Directly using the Zig package, pinned to any version. Nix is "known as the most up to date distribution", but this project uses a *nightly* build of Zig, so that's not there.
+	# - Building Zig from source. The latest version of Zig itself uses LLVM+Clang 15, which is brand new and also not yet in Nix repos. There was a brief window in late August 2022 where Zig could be built with LLVM 14, but the most recent commit that works with that doesn't work with this project.
+	# - Building LLVM 15 from source. I didn't even try this given the massive disk space and build time requirements.	
+	getZigUrl = version: with builtins; ((fromJSON (readFile (fetchurl {
+		# Officially this information is published without past history at: https://ziglang.org/download/index.json
+		# TODO: To switch to a newer version of Zig:
+		# 1. Update this URL with the latest commit.
+		# 2. Try to use it with `nix-shell`. It will error because the hash is wrong, and the right hash will be reported. Put the new hash here too. (I swear this is what the tutorials say to do.)
+		# 3. Change the "YYYY-MM-DD" key below to the new version, as long as it's in the `sources.json`.
+		url = "https://raw.githubusercontent.com/mitchellh/zig-overlay/17352071583eda4be43fa2a312f6e061326374f7/sources.json";
+		sha256 = "sha256:0356h1aaknnzzk0lrjjbg5lwy9b3i06rsww91cmjs45scy6h7fj2";
+	}))).master.${version} or (abort "Unknown version: ${version}")
+	).${pkgs.stdenv.system} or (abort "Unknown platform ${pkgs.stdenv.system}");
+	
+	zig-pinned-bin = let zigUrl=getZigUrl "2022-11-29"; in pkgs.stdenv.mkDerivation {
+		pname = "zig-bin";
+		inherit (zigUrl) version;
+		
+		src = pkgs.fetchurl {
+			inherit (zigUrl) url sha256;
+		};
+		
+		installPhase = ''
+			mkdir -p $out/
+			for _f in *; do
+				echo "Installing file: $_f"
+				cp -r $_f $out/$_f
+			done
+			mkdir -p $out/bin/
+			ln -s $out/zig $out/bin/zig
+		'';
+	};
```

Comment:
If I'm understanding correctly this seems to be recreating the logic used in `zig-overlay`; is there a reason you aren't using the outputs from `zig-overlay` here?

I think you should be able to fetch and import `default.nix` from that repo here (I'm assuming we're okay with IFD).

Reply:
Honestly, because I'm still quite new to this, and thinking in terms of "What would I do if this were a shell script" was faster than reading much into what `zig-overlay` does and how to use it 😅.

Aside from that, I do have some practical concerns about using user GH repos as configuration input (E.G. `zig-overlay`'s README already mentions a previous version of the repo disappearing), convenience (Would it require the user to set an "--experimental" flag?), and stability/support (E.G. Would it be a problem that Debian's stuck at Nix 2.3.7?).

Mostly this comes from a place of inexperience and unfamiliarity, though, so if you say it's fine then we may as well do it.